Error.isError() should also support DOMException

tetsuharuohzeki opened the pull request:
#### 593c460c8c56bb963f298d57775e187acee646a7
<pre>
Error.isError() should also support DOMException
<a href="https://bugs.webkit.org/show_bug.cgi?id=292727">https://bugs.webkit.org/show_bug.cgi?id=292727</a>

Reviewed by NOBODY (OOPS!).

This change is hacky fix for the compatibility with other UAs for `Error.isError()` has been shipped in Safari 18.4.
Ideally, we should make `WebCore::JSDOMException` inherits `JSC::ErrorInstance` properly
but it requires more complicated and deeply change.
This patch postpones them and focuses to improve the compatibility.

* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/es-exceptions/DOMException-is-error.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/es-exceptions/DOMException-is-error.any.worker-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
* Source/WebCore/bindings/scripts/test/JS/JSTestException.h:
(WebCore::JSTestException::createStructure):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/593c460c8c56bb963f298d57775e187acee646a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23313 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108806 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54277 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105670 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31863 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78748 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18363 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93488 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59083 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18173 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11535 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53631 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87956 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11594 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111184 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22699 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87746 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31138 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89688 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87390 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32268 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9992 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25132 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30705 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36017 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30505 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33836 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -3259,6 +3259,8 @@ sub GenerateHeader
         push(@headerContent, "        return JSC::Structure::create(vm, globalObject, prototype, JSC::TypeInfo(JSC::JSType($type), StructureFlags), info(), $indexingModeIncludingHistory);\n");
     } elsif ($codeGenerator->InheritsInterface($interface, "Event")) {
         push(@headerContent, "        return JSC::Structure::create(vm, globalObject, prototype, JSC::TypeInfo(JSC::JSType(JSEventType), StructureFlags), info(), $indexingModeIncludingHistory);\n");
+    } elsif ($interface->extendedAttributes->{Exception}) {
```
ErrorInstanceType's object needs to inherit ErrorInstance.

tetsuharuohzeki replied:
@Constellation 

`JSC::JSWebAssemblyException` also uses `ErrorInstanceType` directly in `JSWebAssemblyException::createStructure` but [`JSC::JSWebAssemblyException`](https://github.com/WebKit/WebKit/blob/154b4e083bbb3a8e1673f88d1e10535bcc65f64f/Source/JavaScriptCore/wasm/js/JSWebAssemblyException.h#L36) does not inherit `JSC::ErrorInstance`.
https://github.com/WebKit/WebKit/blob/154b4e083bbb3a8e1673f88d1e10535bcc65f64f/Source/JavaScriptCore/wasm/js/JSWebAssemblyException.cpp#L47

Is what you mean that ErrorInstanceType's object in JSC's user code (`WebCore::JSDOMException` here) needs to be inherit `JSC::ErrorInstance`?

Constellation replied:
Yes. I think we should change `JSWebAssemblyException` as well. Since these things needs to be handled well in serialized script value's Error serialization as well too. So this patch requires more complicated deep changes.

tetsuharuohzeki replied:
@Constellation

Thanks for your response.

Ideally, I understand we should change `WebCore::JSDOMException` as your suggestion..., but can we go this change as a simple fix due to that [`Error.isError()` has been shipped since Safari 18.4](https://developer.apple.com/documentation/safari-release-notes/safari-18_4-release-notes) and separate a patch into make `WebCore::JSDOMException` inherits `JSC:ErrorInstance` in another bug?

As a compatibility, It's inconvinient that it's not usable `Error.isError(new DOMException) === false`. 

> Since these things needs to be handled well in serialized script value's Error serialization as well too. 

By walk-though our codes, I seem that has been handled in the current class design of `WebCore::JSDOMException` (e.g. 
https://github.com/WebKit/WebKit/blob/aa67d9dda49eac5dd94c0daa5e56013e4c526957/Source/WebCore/bindings/js/JSDOMExceptionHandling.cpp). And I also seem that `ErrorInstanceType` is related to only `JSC::JSObject::isErrorInstance()` so I think this change (without changing the class hierarchy of `WebCore::JSDOMException`) would not be a problem.

Constellation replied:
I don't think it is fine. Using ErrorInstanceType for non ErrorInstance will introduce a new bug and making the code worse.

Constellation replied:
`ErrorInstanceType` is as name said, it must be used for ErrorInstance. JSWebAssemblyException should be fixed, but applying this to the other type is the reversed course, and it will cause the future security issues since we will strongly assume that `ErrorInstanceType` will be only used for derived classes of `ErrorInstance`.

tetsuharuohzeki committed:
Error.isError() should also support DOMException
https://bugs.webkit.org/show_bug.cgi?id=292727

Reviewed by NOBODY (OOPS!).

This change is hacky fix for the compatibility with other UAs for `Error.isError()` has been shipped in Safari 18.4.
Ideally, we should make `WebCore::JSDOMException` inherits `JSC::ErrorInstance` properly
but it requires more complicated and deeply change.
This patch postpones them and focuses to improve the compatibility.

* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/es-exceptions/DOMException-is-error.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebIDL/ecmascript-binding/es-exceptions/DOMException-is-error.any.worker-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateHeader):
* Source/WebCore/bindings/scripts/test/JS/JSTestException.h:
(WebCore::JSTestException::createStructure):

Constellation reviewed (changes requested):
I don't think this is the right fix. It is more error prone for the security bugs.
